### PR TITLE
Backport https://github.com/apache/incubator-beam/pull/918

### PR DESCRIPTION
**Pull request by lukecwik**
R: @peihe

### CODE REVIEW

**Comment by peihe** [comment]
LGTM

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java [hunk 1]
```diff
@@ -613,11 +647,41 @@ private Bound(String name, String filenamePrefix, String filenameSuffix, Coder<T
...
+       * Returns a transform for writing to text files that adds a header string to the files
+       * it writes.
+       *
+       * <p>A {@code null} value will clear any previously configured header.
+       *
+       * <p>Does not modify this object.
+       *
+       * @param header the string to be added as file header
+       */
+      public Bound<T> withHeader(String header) {
```
Comment (dhalperi) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java]: needs to be `@Nullable`

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java [hunk 2]
```diff
@@ -613,11 +647,41 @@ private Bound(String name, String filenamePrefix, String filenameSuffix, Coder<T
...
+       * Returns a transform for writing to text files that adds a footer string to the files
+       * it writes.
+       *
+       * <p>A {@code null} value will clear any previously configured footer.
+       *
+       * <p>Does not modify this object.
+       *
+       * @param footer the string to be added as file footer
+       */
+      public Bound<T> withFooter(String footer) {
```
Comment (dhalperi) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java]: `@Nullable`

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java [hunk 3]
```diff
@@ -698,6 +767,14 @@ public String getFilenameSuffix() {
         return coder;
       }
 
+      public String getHeader() {
```
Comment (dhalperi) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java]: `@Nullable`

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java [hunk 4]
```diff
@@ -698,6 +767,14 @@ public String getFilenameSuffix() {
...
+        return header;
+      }
+
+      public String getFooter() {
```
Comment (dhalperi) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java]: `@Nullable`

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java [hunk 5]
```diff
@@ -976,17 +1053,22 @@ private boolean tryToEnsureNumberOfBytesInBuffer(int minCapacity) throws IOExcep
   @VisibleForTesting
   static class TextSink<T> extends FileBasedSink<T> {
     private final Coder<T> coder;
+    private final String header;
+    private final String footer;
```
Comment (dhalperi) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java]: `@Nullable` both

Comment (dhalperi) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java]: ... stopping commenting here.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java [hunk 6]
```diff
@@ -1014,19 +1100,49 @@ private TextWriteOperation(TextSink<T> sink, Coder<T> coder) {
     private static class TextWriter<T> extends FileBasedWriter<T> {
       private static final byte[] NEWLINE = "\n".getBytes(StandardCharsets.UTF_8);
       private final Coder<T> coder;
+      private final String header;
+      private final String footer;
       private OutputStream out;
 
       public TextWriter(FileBasedWriteOperation<T> writeOperation, Coder<T> coder) {
```
Comment (dhalperi) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java]: why do we need multiple constructors for this private class? Should just have one.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java [hunk 7]
```diff
@@ -1014,19 +1100,49 @@ private TextWriteOperation(TextSink<T> sink, Coder<T> coder) {
...
+      public TextWriter(FileBasedWriteOperation<T> writeOperation, Coder<T> coder, String header,
+                        String footer) {
         super(writeOperation);
+        this.header = header;
+        this.footer = footer;
         this.mimeType = MimeTypes.TEXT;
         this.coder = coder;
       }
 
+      private void writeLine(String line) throws IOException {
```
Comment (dhalperi) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java]: `writeLine` seems a strange name for this function -- why do we think the `String` is only a line?

Javadoc should specify that `null` values will be ignored, but if the value is non-`null` then a newline will be added and the line will not be escaped.

File: sdk/src/test/java/com/google/cloud/dataflow/sdk/io/TextIOTest.java [hunk 8]
```diff
@@ -284,28 +307,71 @@ public void testReadDisplayData() {
...
           }
-          actual.add(line);
+          currentFile.add(line);
         }
+        actual.add(currentFile);
       }
     }
 
-    String[] expected = new String[elems.length];
+    LinkedList<String> expectedElements = Lists.newLinkedList();
```
Comment (dhalperi) [inline sdk/src/test/java/com/google/cloud/dataflow/sdk/io/TextIOTest.java]: Does this need to explicitly be typed as a `LinkedList`? Did not see that used anywhere.

Should maintain consistent style in constructing lists (L310 vs L313) and types if possible (We use `ArrayList` twice and LinkedList once).

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java [hunk 9]
```diff
@@ -1014,19 +1100,49 @@ private TextWriteOperation(TextSink<T> sink, Coder<T> coder) {
...
+          out.write(NEWLINE);
+        }
+      }
+
       @Override
       protected void prepareWrite(WritableByteChannel channel) throws Exception {
         out = Channels.newOutputStream(channel);
       }
 
+     @Override
```
Comment (dhalperi) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java]: mis-aligned spacing?

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java [hunk 10]
```diff
@@ -471,7 +471,31 @@ private Read() {}
...
     }
 
-    // TODO: appendingNewlines, header, footer, etc.
+    /**
+     * Returns a transform for writing to text files that adds a header string to the files
+     * it writes.
+     *
+     * <p>A {@code null} value will clear any previously configured header.
+     *
+     * @param header the string to be added as file header
```
Comment (dhalperi) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/io/TextIO.java]: javadoc should specify that a newline will be unconditionally added after the header and footer.

**Pull request merged**